### PR TITLE
Properly upgrade Casks with version :latest

### PR DESCRIPTION
**Pull request by amyspark**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

When I proposed #3396, I forgot to test if the upgrade process worked for `version :latest` Casks.
Obviously, this came back to bite me 😢 

In this initial state, I've submitted a test to summarize the issue. The missing part would be to tamper with `staged_path` if `version :latest` and use a randomized? path, but I'd like to request your comments on the best way to do it.

This pull request, once finished, fixes caskroom/homebrew-cask#41373.

Thanks for your comments!

/cc @MikeMcQuaid @vitorgalvao @reitermarkus

### CODE REVIEW

**Comment by reitermarkus** [comment]
> Obviously, this came back to bite me 😢

No worries, there had to be *something*. 😜

> and use a randomized? path

`brew reinstall` appends `.reinstall` to the Keg path: https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cmd/reinstall.rb#L74-L76

I'd suggest to just move `staged_path` to `"#{staged_path}.backup"` (or `.upgrading`), and either remove it after the upgrade, or move it back if the upgrade fails.

File: Library/Homebrew/cask/lib/hbc/installer.rb [hunk 1]
```diff
@@ -420,6 +436,14 @@ def zap
       purge_caskroom_path
     end
 
+    def backup_path(path)
+      Pathname.new "#{path}.upgrading" unless path.nil?
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/installer.rb]: When would `path` be `nil`?

File: Library/Homebrew/cask/lib/hbc/installer.rb [hunk 2]
```diff
@@ -378,10 +378,26 @@ def start_upgrade
 
       disable_accessibility_access
       uninstall_artifacts
+      backup if version_is_latest?
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/installer.rb]: Let's always create a backup, not just when `version_is_latest?`.

File: Library/Homebrew/cask/lib/hbc/installer.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/cask/lib/hbc/installer.rb]: Also, since this is only used for `staged_path`, you could make this take no arguments and just return `Pathname("#{@cask.staged_path}.upgrading")`.

Comment (amyspark) [reply Library/Homebrew/cask/lib/hbc/installer.rb]: `path` should never be  `nil`, but I wanted to preserve the semantics of `gain_permissions_remove(staged_path) if !staged_path.nil? && staged_path.exist?`.

File: Library/Homebrew/cask/lib/hbc/installer.rb (hunk 2 above)
Comment (amyspark) [reply Library/Homebrew/cask/lib/hbc/installer.rb]: If we always create a backup, we'll need a way to know if this is the new or old cask when `purge_versioned_files`. Otherwise, it will leave a dangling, empty folder if the upgrade fails (see the test `"restores the old Cask if the upgrade failed"` in `upgrade_spec.rb`)

Comment (reitermarkus) [reply Library/Homebrew/cask/lib/hbc/installer.rb]: Isn't `start_upgrade` only called on the old cask?

Comment (amyspark) [reply Library/Homebrew/cask/lib/hbc/installer.rb]: Yes, I took advantage of that; see next commit.

File: Library/Homebrew/cask/lib/hbc/installer.rb [hunk 3]
```diff
@@ -420,6 +437,19 @@ def zap
       purge_caskroom_path
     end
 
+    def backup_path
+      return nil if @cask.staged_path.nil?
+      if backed_up?
+        Pathname.new "#{@cask.staged_path}.upgrading"
+      else
+        @cask.staged_path
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/installer.rb]: I think this is really hard to follow. The only reason for this is `purge_versioned_files` using it. I'd suggest making two specialised versions of `purge_versioned_files`, one normal one, and one for the `upgrade?` case.

Comment (amyspark) [reply Library/Homebrew/cask/lib/hbc/installer.rb]: I'll do it that way, and then factor out the whole metadata handling (to avoid code duplication).

File: Library/Homebrew/cask/lib/hbc/installer.rb [hunk 4]
```diff
@@ -420,19 +434,43 @@ def zap
...
       ohai "Purging files for version #{@cask.version} of Cask #{@cask}"
 
       # versioned staged distribution
       gain_permissions_remove(@cask.staged_path) if !@cask.staged_path.nil? && @cask.staged_path.exist?
 
       # Homebrew-Cask metadata
+      purge_metadata
+    end
+
+    def purge_metadata
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/installer.rb]: The lines in here which have `unless upgrade?` could be moved to `purge_versioned_files` now.

File: Library/Homebrew/cask/lib/hbc/installer.rb [hunk 5]
```diff
@@ -420,19 +434,43 @@ def zap
...
       if @cask.metadata_versioned_path.respond_to?(:children) &&
-         @cask.metadata_versioned_path.exist?
+         @cask.metadata_versioned_path.exist? &&
+         !(upgrade? && version_is_latest?)
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/installer.rb]: Do we still need `version_is_latest?` now?

File: Library/Homebrew/cask/lib/hbc/installer.rb [hunk 6]
```diff
@@ -420,19 +434,43 @@ def zap
...
+
+    def version_is_latest?
+      @cask.versions.include?("latest")
+    end
+
     def gain_permissions_remove(path)
       Utils.gain_permissions_remove(path, command: @command)
     end
 
+    def purge_backed_versioned_files
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/installer.rb]: `purge_backed_up_versioned_files`

File: Library/Homebrew/cask/lib/hbc/installer.rb (hunk 5 above)
Comment (amyspark) [reply Library/Homebrew/cask/lib/hbc/installer.rb]: Yes, the backup mechanism doesn't involve the Cask's metadata; if the version is `latest`, deleting this line will make `purge_metadata` "uninstall" the Cask while leaving all artifacts in place.

Comment (amyspark) [reply Library/Homebrew/cask/lib/hbc/installer.rb]: On second thoughts, I'll just go ahead with two fully separate methods and get rid of that line.

File: Library/Homebrew/cask/lib/hbc/installer.rb [hunk 7]
```diff
@@ -420,10 +437,46 @@ def zap
...
+      return nil if @cask.staged_path.nil?
+      Pathname.new "#{@cask.staged_path}.upgrading"
+    end
+
+    def backup_metadata_path
+      return nil if @cask.metadata_versioned_path.nil?
+      Pathname.new "#{@cask.metadata_versioned_path}.upgrading"
+    end
+
+    def version_is_latest?
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/installer.rb]: This can now be removed.

File: Library/Homebrew/cask/lib/hbc/installer.rb [hunk 8]
```diff
@@ -420,10 +437,46 @@ def zap
...
+      if backup_metadata_path.respond_to?(:children) &&
+         backup_metadata_path.exist? &&
+         backup_metadata_path.children.each do |subdir|
+           unless PERSISTENT_METADATA_SUBDIRS.include?(subdir.basename)
+             gain_permissions_remove(subdir)
+           end
+         end
+      end
+      backup_metadata_path.rmdir_if_possible
+      @cask.metadata_master_container_path.rmdir_if_possible unless upgrade?
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/installer.rb]: The `unless upgrade?` here doesn't make sense, because this method is only called when upgrading.

File: Library/Homebrew/cask/lib/hbc/installer.rb [hunk 9]
```diff
@@ -31,7 +31,7 @@ def initialize(cask, command: SystemCommand, force: false, skip_cask_deps: false
       @upgrade = upgrade
     end
 
-    attr_predicate :binaries?, :force?, :skip_cask_deps?, :require_sha?, :upgrade?, :verbose?
+    attr_predicate :binaries?, :force?, :skip_cask_deps?, :require_sha?, :upgrade?, :verbose?, :backed_up?
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/installer.rb]: This is not needed anymore.

File: Library/Homebrew/cask/lib/hbc/artifact/moved.rb [hunk 10]
```diff
@@ -71,7 +71,7 @@ def move_back(skip: false, force: false, command: nil, **options)
           command.run("/bin/mv", args: [target, source], sudo: true)
         end
 
-        add_altname_metadata(target, source.basename, command: command)
+        add_altname_metadata(source, target.basename, command: command)
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/artifact/moved.rb]: Remove this completely. We don‘t need to add this when moving it back, actually.

File: Library/Homebrew/cask/lib/hbc/installer.rb [hunk 11]
```diff
@@ -420,10 +437,38 @@ def zap
...
 
+    def purge_backed_up_versioned_files
+      ohai "Purging files for version #{@cask.version} of Cask #{@cask}"
+
+      # versioned staged distribution
+      gain_permissions_remove(backup_path) if !backup_path.nil? && backup_path.exist?
+
+      # Homebrew-Cask metadata
+      if backup_metadata_path.respond_to?(:children) &&
+         backup_metadata_path.exist? &&
```
Comment (reitermarkus) [inline Library/Homebrew/cask/lib/hbc/installer.rb]: The `&&` shouldn‘t be here.

Comment (reitermarkus) [reply Library/Homebrew/cask/lib/hbc/installer.rb]: The whole condition here could probably just be `backup_metadata_path.directory?`.

**Comment by reitermarkus** [comment]
You probably missed this comment by a minute: https://github.com/Homebrew/brew/pull/3492#discussion_r153669179

**Comment by amyspark** [comment]
Yep, I wasn't using my inbox to keep up with comments. Thanks for the heads up @reitermarkus.

**Comment by reitermarkus** [comment]
Thanks again, @amyspark! 👍
